k2/csrc/tensor.cu

```diff
@@ -131,9 +132,57 @@ void Tensor::Init(ContextPtr c) {
 }
 
 Tensor ToContiguous(const Tensor &tensor) {
-  // TODO(haowen): implement
-  K2_LOG(FATAL) << "Not implemented";
-  return tensor;
+  if (tensor.IsContiguous()) return tensor;
+  std::vector<int32_t> dims = tensor.Dims();
+  std::vector<int32_t> strides = tensor.Strides();
+  int32_t num_dims = static_cast<int32_t>(dims.size());
+
+  std::size_t num_dims_bytes = num_dims * sizeof(int32_t);
+
+  RegionPtr dims_ptr = NewRegion(tensor.Context(), num_dims_bytes);
+  RegionPtr strides_ptr = NewRegion(tensor.Context(), num_dims_bytes);
+
+  auto cpu_context = tensor.Context()->GetCpuContext();
+  auto kind = GetMemoryCopyKind(*cpu_context, *tensor.Context());
+  MemoryCopy(dims_ptr->data, static_cast<void *>(dims.data()), num_dims_bytes,
+             kind);
+  MemoryCopy(strides_ptr->data, static_cast<void *>(strides.data()),
+             num_dims_bytes, kind);
+
+  const int32_t *dims_data = dims_ptr->GetData<int32_t>();
+  const int32_t *strides_data = strides_ptr->GetData<int32_t>();
+
+  // turn a linear index of a multi-dimensional array
+  // first into a coordinate representation, and then to offset
+  // in number of elements via strides information.
+  auto to_offset = [=] __host__ __device__(int32_t i) -> int32_t {
+    int32_t offset = 0;
+    for (int32_t k = num_dims - 1; k > 0; --k) {
+      int32_t index = i % dims_data[k];
+      offset += index * strides_data[k];
+      i = (i - index) / dims_data[k];
+      if (i == 0) break;
+    }
+    offset += i * strides_data[0];
+    return offset;
+  };
+
+  Tensor ans(tensor.Context(), tensor.GetDtype(), dims);
+
+  char *ans_data = static_cast<char *>(ans.Data());
+  const char *this_data = static_cast<const char *>(tensor.Data());
+  int32_t element_size = tensor.ElementSize();
+
+  auto copy_data = [=] __host__ __device__(int32_t i) -> void {
+    char *dst = ans_data + i * element_size;
+    const char *src = this_data + to_offset(i) * element_size;
```

Comment:
This is super inefficient, doing things character by character.   Better to have this templated on the dtype.

Reply:
... or at least we can have versions of the kernel that are 1-byte, 2-byte, 4-byte and 8-byte, and we select the largest one that's a multiple of the dtype's number of characters (that can be looked up from the type traits).

Reply:
... and my preference would be to handle num_axes of (0,1) and (2) separately, using Eval() and Eval2().  That will be much more efficient than this.  Sorry, I know it will increase the amount of code that has to be written.

Reply:
Can we just implement cases `num_axes = 0,1,2,3` for now as we may only need those versions for now (add version num_axes>=4 if we really need)? As Cuda supports `dim3` for `block/thread`, we can implement `1,2,3` version with `Eval`, `Eval2`, `Eval3` straightforward.

Reply:
Replaced with Dan's implementation